### ISSUE TITLE
Refactor project panel

### ISSUE THREAD

**Issue by thomaswilburn** [thomaswilburn/Caret#252]
These are some notes on plans to refactor projects so that they suck less. I'm not at all happy with the way that the current project file tree is retained and walked, but I'm also not sure that the packaged app API provides a strong enough base for rebuilding on a cleaner foundation--it's a weird mix of HTML5 file system APIs and the Chrome file entry objects, the former of which is now apparently deprecated(!) anyway.

Work on this will proceed in the WeakMaps branch, so that I can throw big chunks of it away if I feel like it.
- Project needs to be split into multiple modules, with a `projects/*` namespace containing the code for walking/monitoring the filesystem, and a separate `ui/projectPanel` module just for view code.
- Once WeakMap support enters Chrome Stable, we can use them to connect DOM elements to filesystem node objects, which will make progressive rendering a lot easier.
- Rather than actually retaining a set of heavy FSNode objects, and in the absence of file watch operations, it may make sense to just walk the tree and track all the paths that are found, using those as keys to entries (if necessary) or just using getFile() to grab them based on path from the directory root (note: in testing, this did not seem to work in Chrome 35, which means we still need to keep thousands of file entries around).
- On the other hand, if filesytem watch APIs are added, nodes will be capable of watching their entries and launching a re-render of their associated DOM node on changes, which would greatly reduce the amount of special-case bookkeeping we have to do manually for updates.
- Hybrid approach: keep the FS Node objects, but do not store a tree, just an array mapping paths to file entries. Use regex to filter the list when performing rendering or update operations. Preliminary testing on the Caret project indicates this could be relatively fast (around 2ms for 2,057 files). The upshot of this is much simpler bookkeeping, and easier updates without the watch API (just use the pathname to find the parent directory and refresh).

**Comment by thomaswilburn** [comment]
Work on this has started in the `preview-rewrite` branch, after `weakmaps` was merged in a while back. The current redesign still retains a tree of node objects, but will render these on-demand instead of trying to do so all in one batch. It'll also start splitting the config and view sides of the project functionality into separate modules.

We are still limited by two factors: we need a complete list of files in order to do in-project search, which means we need to spend a bunch of time running around in the file system, and we cannot move this work to another thread because the Chrome API is not available from a web worker. As far as I can tell, there's no plan to change this in Chrome (a similar discussion existed for service workers and ended in the same desultory whimper).

So the result of this is that projects will be cleaner, which is good, because currently the projectManager.js file is a huge, ugly mess (by far our most hideous module). But it isn't likely to get much faster until we either have a way to move this off the main thread, or there's a better API for interacting with the filesystem.

@benfredwells I know at some point, you had a pull request for a rewrite of the panel rendering that I didn't take, and you're closer to the source than I am. Do you know if there are any plans for the filesystem API to get revisited in the future? It could just be me reading the wrong tea leaves, but it seems like between ServiceWorker and the Add to Homescreen improvements, whatever plans existed for manifest v3 and a Chrome Apps revision have been put on hold.

**Comment by benfredwells** [comment]
There are no plans to do anything substantial with the filesystem API. I don't know what is happening with manifest v3, but you are right that a lot of our focus has shifted to ServiceWorker and Add to Homescreen (the latter should be coming to desktop btw).

We don't have any concrete plans right now to add any file system capabilities to ServiceWorker based apps but it is something we're actively discussing.

**Comment by thomaswilburn** [comment]
Checking in on this: Basic functionality seems to work. Directories can be removed. Rendering is lazy, which is nice.

Currently untested: project files, including integration into settings. Once this is done, I think we'll be good to go.

Current ETA: maybe done by the time this quarter ends, so I can hotpatch the inevitable bugs.
